### PR TITLE
FEATURE: provide more details when performing a bulk add to group

### PR DESCRIPTION
**Pull request by techAPJ**
https://meta.discourse.org/t/provide-more-details-when-performing-a-bulk-add-to-group/60665?u=techapj

This PR adds more information on success page.

screen shot 2017-04-27 at 01 25 44

### CODE REVIEW

File: app/controllers/admin/groups_controller.rb [hunk 1]
```diff
@@ -24,13 +24,25 @@ def bulk
...
+      end
+      invalid_users = users.reject! { |u| valid_emails[u] || valid_usernames[u] }
+      valid_users.flatten!
+      group.bulk_add(valid_users) if valid_users.present?
+      users_added = valid_users.count
     end
 
-    render json: success_json
+    message = I18n.t('groups.success.bulk_add', users_added: users_added)
+    message = "#{message} <br><br> #{I18n.t('groups.success.bulk_add_invalid_users', invalid_users: invalid_users.join(', '))}" unless invalid_users.blank?
```
Comment (eviltrout) [inline app/controllers/admin/groups_controller.rb]: I am not a huge fan of building HTML on the server side like this.

Why not return the invalid users as an array that you could display in the hbs template by using an #each?

Comment (techAPJ) [inline app/controllers/admin/groups_controller.rb]: Good point. Made it so.

File: app/controllers/admin/groups_controller.rb [hunk 2]
```diff
@@ -24,13 +24,22 @@ def bulk
 
   def bulk_perform
     group = Group.find(params[:group_id].to_i)
+    users_added = 0
     if group.present?
       users = (params[:users] || []).map {|u| u.downcase}
-      user_ids = User.where("username_lower in (:users) OR email IN (:users)", users: users).pluck(:id)
-      group.bulk_add(user_ids) if user_ids.present?
+      valid_emails = valid_usernames = {}
```
Comment (tgxworld) [inline app/controllers/admin/groups_controller.rb]: Both of these variables are actually referring to the same hash so I'm not sure if there is what you intended?

File: app/controllers/admin/groups_controller.rb [hunk 3]
```diff
@@ -24,13 +24,22 @@ def bulk
...
+      valid_users = User.where("username_lower IN (:users) OR email IN (:users)", users: users).pluck(:id, :username_lower, :email)
+      valid_users.each do |vu|
```
Comment (techAPJ) [inline app/controllers/admin/groups_controller.rb]: I am modifying the array as well via `vu.slice!(1..2)`.. (this is done so as to pass IDs only to `bulk_add` method).

File: app/controllers/admin/groups_controller.rb (hunk 2 above)
Comment (techAPJ) [inline app/controllers/admin/groups_controller.rb]: Oops, yeah. Fixed now.

File: app/controllers/admin/groups_controller.rb (hunk 3 above)
Comment (tgxworld) [inline app/controllers/admin/groups_controller.rb]: Ah in that case, you can do the following which I feel is much easier to understand and read.

```
valid_users.map! do |id, username_lower, email| 
  valid_emails[email] =  valid_usernames[username_lower] = id
  id
```

Comment (techAPJ) [inline app/controllers/admin/groups_controller.rb]: Awesome, thanks for the tip! Done via https://github.com/discourse/discourse/commit/e3f82140d8fa7f7e01378ccd5dcad57a96f2909b.

**Pull request merged**